tests/Builder/BuiderCodecTest.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace MongoDB\Tests\Builder;
+
+use Generator;
+use MongoDB\Builder\Aggregation;
+use MongoDB\Builder\BuilderCodec;
+use MongoDB\Builder\Expression;
+use MongoDB\Builder\Pipeline;
+use MongoDB\Builder\Query;
+use MongoDB\Builder\Stage;
+use MongoDB\Tests\TestCase;
+
+use function array_merge;
+
+class BuiderCodecTest extends TestCase
+{
+    public function testPipeline(): void
+    {
+        $pipeline = new Pipeline(
+            Stage::match(Aggregation::eq('$foo', 1)),
```

Comment:
Note that the `$match` stage shouldn't accept the aggregation pipeline operators, except within `$expr`.